src/main/java/org/opentripplanner/transit/service/StopModel.java

```diff
@@ -36,293 +29,246 @@ public class StopModel implements Serializable {
-  private final Map<FeedScopedId, GroupOfStations> groupOfStationsById = new HashMap<>();
-
-  private final Map<FeedScopedId, TransitStopVertex> transitStopVertices = new HashMap<>();
+  private final Map<FeedScopedId, Stop> regularStopsById;
+  private final Map<FeedScopedId, Station> stationById;
+  private final Map<FeedScopedId, MultiModalStation> multiModalStationById;
+  private final Map<FeedScopedId, GroupOfStations> groupOfStationsById;
+  private final Map<FeedScopedId, FlexStopLocation> flexStopsById;
+  private final Map<FeedScopedId, FlexLocationGroup> flexStopGroupsById;
 
   /** The density center of the graph for determining the initial geographic extent in the client. */
-  private Coordinate center = null;
-
-  private final Map<FeedScopedId, FlexStopLocation> locationsById = new HashMap<>();
-
-  private final Map<FeedScopedId, FlexLocationGroup> locationGroupsById = new HashMap<>();
+  private final WgsCoordinate stopLocationCenter;
 
   private transient StopModelIndex index;
 
   @Inject
-  public StopModel() {}
-
-  public void index() {
-    LOG.info("Index stop model...");
-    index = new StopModelIndex(this);
-    LOG.info("Index stop model complete.");
+  public StopModel() {
+    this.regularStopsById = Map.of();
+    this.stationById = Map.of();
+    this.multiModalStationById = Map.of();
+    this.groupOfStationsById = Map.of();
+    this.flexStopsById = Map.of();
+    this.flexStopGroupsById = Map.of();
+    this.stopLocationCenter = null;
+    this.index = new StopModelIndex(List.of(), List.of(), List.of(), List.of());
   }
 
-  private StopModelIndex getStopModelIndex() {
-    //TODO refactoring transit model - thread safety
-    if (index == null) {
-      index();
-    }
-    return index;
+  public StopModel(StopModelBuilder builder) {
+    this.regularStopsById = builder.stopsById().asImmutableMap();
+    this.stationById = builder.stationsById().asImmutableMap();
+    this.multiModalStationById = builder.multiModalStationsById().asImmutableMap();
+    this.groupOfStationsById = builder.groupOfStationsById().asImmutableMap();
+    this.flexStopsById = builder.flexStopsById().asImmutableMap();
+    this.flexStopGroupsById = builder.flexStopGroupsById().asImmutableMap();
+    this.stopLocationCenter = builder.calculateTransitCenter();
+    reindex();
   }
 
-  /**
-   * @param id Id of Stop, Station, MultiModalStation or GroupOfStations
-   * @return The associated TransitStopVertex or all underlying TransitStopVertices
-   */
-  @Nullable
-  public Set<Vertex> getStopVerticesById(FeedScopedId id) {
-    var stops = getStopsForId(id);
-
-    if (stops == null) {
-      return null;
-    }
+  public static StopModelBuilder of() {
+    return new StopModelBuilder();
+  }
 
-    return stops
-      .stream()
-      .filter(Stop.class::isInstance)
-      .map(Stop.class::cast)
-      .map(it -> getStopModelIndex().getStopVertexForStop(it))
-      .collect(Collectors.toSet());
+  public StopModelBuilder copy() {
```

Comment:
Sometimes we have used `mutate()` for methods like these and sometimes `copy()`. Which do you think is the one we should use in the future?

Reply:
Ahh, very good point - both works for me, mutate() is probably better considering the return value.